### ISSUE TITLE
OMF commands saying: `set_color: Unknown color '^'`

### ISSUE THREAD

**Issue by cableray** [oh-my-fish/oh-my-fish#861]
I'm getting issues with OMF commands, including `omf doctor`, returning `set_color: Unknown color '^'` (twice in a row).

This seems to be happening with most OMF commands, and I suspect it may be an incompatibility with fish 3.3.1 and fixed in OMF 7, but `omf update` is plagued with this issue and doesn't seem to be working...

```
omf update omf
fatal: bad revision '^'
set_color: Unknown color '^'
set_color: Unknown color '^'

set_color: Unknown color '^'
set_color: Unknown color '^'

set_color: Unknown color '^'
set_color: Unknown color '^'

Updating https://github.com/oh-my-fish/packages-main master... Done!
```

Is there a way to do a manual update? I can try that first....

**Environment report**

```
Oh My Fish version:   6
OS type:              Darwin
Fish version:         fish, version 3.3.1
Git version:          git version 2.33.0
Git core.autocrlf:    input
set_color: Unknown color '^'
set_color: Unknown color '^'
```

**Comment by scorphus** [comment]
Yeah, I saw this coming. Some old version of Oh My Fish encounters a new version of Fish. Not matter how [loud that could have got telegraphed](https://github.com/fish-shell/fish-shell/issues/7105#issuecomment-645171108), this would happen regardlessly.

So, your options:

1. Update Oh My Fish manually
```fish
git -C $HOME/.local/share/omf remote update
git -C $HOME/.local/share/omf reset origin/master --hard
omf reload
```
2. Downgrade Fish, update Oh My Fish, upgrade Fish
```
# That depends on your setup
```

Do you see any other option, @faho?

**Comment by rinkimekari** [comment]
this is now happening to me after doing a system upgrade. I'm using the agnoster theme and its saying `set_color: Unknown color '~'` whenever I start the fish shell. i tried the fixes you guys suggested but its still broken. Any tips?

**Comment by Yofou** [comment]
Yeah me to @rinkimekari, tried everything I found online nothing works, other than doing `omf reload` everytime I open up the terminal

**Comment by scorphus** [comment]
@rinkimekari it only happend when you start the shell?

What theme are you using, @Yofou?

**Comment by rinkimekari** [comment]
> @rinkimekari it only happend when you start the shell?

It's whenever i get the prompt. So it happens when starting the shell, and after executing a command.

**Comment by scorphus** [comment]
Could you please post the contents of both `fish_prompt` and `fish_right_prompt` functions?

You can use `type` for that, e.g.:

```
type fish_prompt > /tmp/fish_prompt.txt
type fish_right_prompt > /tmp/fish_right_prompt.txt
```

⚠️ Please, caution not to post any sensitive information that may be present in these contents! ⚠️

After you post them here, you can delete those files.

**Comment by Yofou** [comment]
@scorphus @rinkimekari, I'm using **agnoster** and I get that error message on startup and on any command I execute (same as rinkimekari), some interesting development is that it's only for the agnoster theme, when I swapped to lambda or default it was perfectly fine (no issues at all) 🤔 

but just in case I'll post the fish_prompts files

### fish_prompt.txt
```fish
fish_prompt is a function with definition
# Defined interactively
function fish_prompt
                _ksi_start_prompt
                _ksi_end_prompt
            
end
```
### fish_right_prompt.txt
```fish
fish_right_prompt is a function with definition
# Defined in /home/yofou/.local/share/omf/themes/agnoster/functions/fish_right_prompt.fish @ line 76
function fish_right_prompt --description 'Prints right prompt'
  if test "$fish_key_bindings" = "fish_vi_key_bindings"
    prompt_vi_mode
    set_color normal
  end
end
```

**Comment by scorphus** [comment]
@Youfou  `fish_right_prompt` funciton definition looks like expected:

```fish
fish_right_prompt is a function with definition
# Defined in /home/nemo/.local/share/omf/themes/agnoster/functions/fish_right_prompt.fish @ line 76
function fish_right_prompt --description 'Prints right prompt'
  if test "$fish_key_bindings" = "fish_vi_key_bindings"
    prompt_vi_mode
    set_color normal
  end
end
```

which indicates that the theme is up to date – especially considering its location `../themes/agnoster/functions/...` which results from the [most recent change in the theme](https://github.com/oh-my-fish/theme-agnoster/pull/53).

`fish_prompt`, however, looks very different from the expected:

```fish
fish_prompt is a function with definition
# Defined in /home/nemo/.local/share/omf/themes/agnoster/functions/fish_prompt.fish @ line 297
function fish_prompt
  set -g RETVAL $status
  prompt_status
  prompt_virtual_env
  prompt_user
  prompt_dir
  if [ (cwd_in_scm_blacklist | wc -c) -eq 0 ]
    type -q hg;  and prompt_hg
    type -q git; and prompt_git
    if [ "$theme_svn_prompt_enabled" = "yes" ]
      type -q svn; and prompt_svn
    end
  end
  prompt_finish
end
```

So that is something to investigate. We're left with a function that reads differently and is defined interactively – i.e. defined by a script and not by a file named `fish_functions.fish` located in a directory that is part of `$fish_function_path`. Namely:

```fish
fish_prompt is a function with definition
# Defined interactively
function fish_prompt
                _ksi_start_prompt
                _ksi_end_prompt
            
end
```

- What does `type _ksi_start_prompt` output?
- What does `type _ksi_end_prompt` output?

Until we get to the cause of the issue, one possible workaround may be re-sourcing `fish_prompt` from agnoster:

```fish
source ~/.local/share/omf/themes/agnoster/functions/fish_prompt.fish
```

**Comment by Yofou** [comment]
@scorphus I really appreciate you helping me out here, just wanted to say that before we continue
here are the outputs

### _ksi_start_prompt
```fish
function _ksi_start_prompt
            set --local cmd_status "$status"
            if test "$_ksi_prompt_state" != "postexec" -a "$_ksi_prompt_state" != "first-run"
                _ksi_mark "D"
            end
            set --global _ksi_prompt_state "prompt_start"
            _ksi_mark "A"
            return "$cmd_status" # preserve the value of $status
        
end
```

### _ksi_end_prompt
```fish
function _ksi_end_prompt
            set --local cmd_status "$status"
            # fish trims one trailing newline from the output of fish_prompt, so
            # we need to do the same. See https://github.com/kovidgoyal/kitty/issues/4032
            set --local op (_ksi_original_fish_prompt) # op is an array because fish splits on newlines in command substitution
            if set -q op[2]
                printf '%s\n' $op[1..-2] # print all but last element of array, each followed by a new line
            end
            printf '%s' $op[-1] # print the last component without a newline
            set --global _ksi_prompt_state "prompt_end"
            _ksi_mark "B"
            return "$cmd_status" # preserve the value of $status
        
end
```

**Comment by scorphus** [comment]
You're welcome, @Yofou.

1. I assume you're using kitty. Correct? I used kitty in the past but these `_ksi_*` functions are new to me. Are they something you activate?

2. What does `type _ksi_original_fish_prompt` tell us?

3. And, does that workaround solves the issue?

   ```fish
   source ~/.local/share/omf/themes/agnoster/functions/fish_prompt.fish
   ```

**Comment by rinkimekari** [comment]
Sorry for such a late reply. Here is my output for fish_prompt:
```
fish_prompt is a function with definition
# Defined in ~/.local/share/omf/themes/agnoster/functions/fish_prompt.fish @ line 297
function fish_prompt
  set -g RETVAL $status
  prompt_status
  prompt_virtual_env
  prompt_user
  prompt_dir
  if [ (cwd_in_scm_blacklist | wc -c) -eq 0 ]
    type -q hg;  and prompt_hg
    type -q git; and prompt_git
    if [ "$theme_svn_prompt_enabled" = "yes" ]
      type -q svn; and prompt_svn
    end
  end
  prompt_finish
end
```
and fish_right_prompt
```
fish_right_prompt is a function with definition
# Defined in ~/.local/share/omf/themes/agnoster/functions/fish_right_prompt.fish @ line 76
function fish_right_prompt --description 'Prints right prompt'
  if test "$fish_key_bindings" = "fish_vi_key_bindings"
    prompt_vi_mode
    set_color normal
  end
end
```
From a quick glance, these seem to be the expected outputs for these functions based on your comment. Not sure why this would be happening lol. I am using kitty as well.

Something else weird, this only happens when my user's shell is set as fish and I open it in a new terminal. If I start fish from another process, it works fine. For example, if I start fish from fish, the inner fish works fine. The same goes for starting from zsh.

Also, I put the workaround you gave in my config.fish and everything works perfectly. Thanks for all the help!

**Comment by Krapaince** [comment]
I have the same issue as @rinkimekari and @Yofou. I use fish 3.3.1 with omf 7 in kitty 0.24.0 and the `agnoster` theme.

I have the same outputs as  @rinkimekari and @Yofou for all of those commands:
```
type fish_prompt
type fish_right_prompt
type _ksi_start_prompt
type _ksi_end_prompt
```
> I assume you're using kitty. Correct? I used kitty in the past but these _ksi_* functions are new to me. Are they something you activate?

Yes, I haven't done any change in my fish config. It seems that it as something to do with the last update of kitty. For instance with kitty `0.23.1` everything works as expected, while updating to the last version `0.24.0` results in `set_color: Unknown color "$[current working directory}":
```
  cd /
set_color: Unknown color “/”
set_color: Unknown color “/”
  cd /home
set_color: Unknown color “/home”
set_color: Unknown color “/home”
  cd /home/krapaince
set_color: Unknown color “~”
set_color: Unknown color “~”
```
> What does type _ksi_original_fish_prompt tell us?
```fish
_ksi_original_fish_prompt is a function with definition
# Defined interactively
function _ksi_original_fish_prompt
  set -g RETVAL $status
  prompt_status
  prompt_virtual_env
  prompt_user
  prompt_dir
  if [ (cwd_in_scm_blacklist | wc -c) -eq 0 ]
    type -q hg;  and prompt_hg
    type -q git; and prompt_git
    if [ "$theme_svn_prompt_enabled" = "yes" ]
      type -q svn; and prompt_svn
    end
  end
  prompt_finish
end
```

> And, does that workaround solves the issue?
 `source ~/.local/share/omf/themes/agnoster/functions/fish_prompt.fish`

Adding this at the end of my config and opening a new kitty result in the same issue (`set_color: Unknown ...`). However, when adding the source and starting fish from an existing shell, it works fine.

What is strange is that over ssh with the same version and config for fish and omf, everything is working as expected.

**Comment by scorphus** [comment]
Thank you all for your inputs. Please check the PR referenced above.

Meanwhile, as a temporary workaround, please add the following to `config.fish`:

```fish
# deactivate kitty shell integration
set -e KITTY_SHELL_INTEGRATION
```

Or, if you prefer not changing that for now, issuing `exec fish -li` once you start a new session might also serve as a temporary workaround. Please LMK if that works.

**Comment by Krapaince** [comment]
> deactivate kitty shell integration
`set -e KITTY_SHELL_INTEGRATION`

Thanks for this temporary work around it works with kitty 0.24.0.

**Comment by tranphuquy19** [comment]
> Simply start fish with the `^` redirections reenabled:
> 
> ```fish
> fish --features no-stderr-nocaret
> ```
> 
> Then do the update.

This actually worked!

**Comment by wudixm** [comment]
Same error happened to me. 

set_color: Unknown color '^'

I tried 
`source ~/.local/share/omf/themes/agnoster/functions/fish_prompt.fish` it works

And another workaround is I changed prompt through

`fish_config` 
and set other prompt and colors, then it works
